include/seastar/http/client.hh

```diff
@@ -38,10 +41,15 @@ namespace experimental {
  * Check the demos/http_client_demo.cc for usage example
  */
 
-class connection {
+class connection : public enable_shared_from_this<connection> {
+    friend class client;
+    using hook_t = bi::list_member_hook<bi::link_mode<bi::auto_unlink>>;
+
     connected_socket _fd;
     input_stream<char> _read_buf;
     output_stream<char> _write_buf;
+    hook_t _hook;
+    seastar::shared_ptr<connection> _pooled; // temporary
```

Comment:
Why both shared_ptr and intrusive lists? You can use boost::intusive_ptr.

Reply:
And shouldn't this be in a subclass? A connection pool makes pooled connections, but the basic_connection_factory makes unpooled connections.

Reply:
> Why both shared_ptr and intrusive lists?

Is your concern is about the `_pooled` member itself? Its temporary, the goal is to keep connection allocated while it's sitting in the list. Next patch removes this field and leaves only the hook on-board.

> And shouldn't this be in a subclass? A connection pool makes pooled connections, but the basic_connection_factory makes unpooled connections.


A connection factory makes `connected_socket`s which are not pooled/cached/whatever, then http client wraps them with `http::connection`s and pools all of them

Reply:
Yes, but why do we have both reference counting and linked lists? If you use intrusive_ptr you don't need shared_ptr.

Reply:
Linked list is to keep connection in pool after client::put_connection(). Reference counting might be a bit of an overkill, the reference counter doesn't really need to grow above 1, but it makes everything significantly simpler. That's how

When the connection is created it does

```c++
connection::connection(connected_socket&& fd)
        : _fd(std::move(fd))
        , _closed(_fd.wait_input_shutdown().finally([me = shared_from_this()] {}))
{
}
```

so the connection references itself by the input-shutdown continuation chain. This implements the "delete the connection when it's put by the caller AND when it's input is closed". I (almost) coded this without shared-ptr, but caught myself re-implementing shared-ptr logic.

Reply:
Ok. If the shared_ptr doesn't leak into the interface, that's fine.